homeassistant/helpers/state.py

```diff
@@ -93,88 +93,90 @@ def get_changed_since(states: Iterable[State],
-        to services
+        Function is provided with a schema for each potential service. This
+        schema is extended with a state entry for services wich take no
+        parameter and modify state like  turn_on/off and such.
         """
         calls = []
 
-        # expanded attribute list containing entity id
-        expanded_attributes = dict(state.attributes)
-        expanded_attributes[ATTR_ENTITY_ID] = state.entity_id
-
-        def add_if_complete(service: str) -> None:
-            """Add a service call with the attributes needed."""
-            schema = vol.Schema(
-                services[service].schema,
-                required=services[service].required,
-                extra=vol.REMOVE_EXTRA)
+        # expanded dict containing entity id and state
+        attributes = dict(state.attributes)
+        attributes[ATTR_ENTITY_ID] = state.entity_id
+        attributes[ATTR_STATE] = state.state
 
+        for schema, service in service_schemas:
             try:
-                attributes = schema(expanded_attributes)
+                validated = schema(attributes)
             except vol.error.Invalid as e:
                 _LOGGER.debug("Incomplete attribute "
                               "list to call {}:{}".format(service, e))
-                return
+                continue
+
+            # state is only syntetic for validation
+            validated.pop(ATTR_STATE, None)
 
             # consume attributes taken by service
             # this avoid multiple service calls
             # for same attribute if schema would
             # allow it
-            for key in attributes:
-                expanded_attributes.pop(key, None)
+            for key in validated:
+                attributes.pop(key, None)
 
             calls.append(
                 ServiceCall(domain,
                             service,
-                            data=attributes,
+                            data=validated,
                             context=state.context))
 
-        # first try to reproce the state value, taking all attributes
-        if state.state in state_to_service:
-            add_if_complete(state_to_service[state.state])
-        elif state.state is None:
-            pass
-        else:
-            _LOGGER.error("Unknown state type {}".format(state.state))
-
-        # next try to reproduce attributes (ordered after)
-        for service in services:
-            if service not in state_to_service.values():
-                add_if_complete(service)
-
         _LOGGER.debug(calls)
         return calls
 
 
+@bind_hass
+def async_construct_service_schemas(
+        hass: HomeAssistantType,
+        domain: str,
+        services: List[Tuple[str, Dict]]) -> Tuple:
+    """Helper function to construct an extended schema for state restore."""
+
+    for service, extension in services:
+        services = hass.services.async_services()[domain]
```

Comment:
Looks like we're modifying the list we're iterating here.

Reply:
Indeed. Odd it worked. Will fix